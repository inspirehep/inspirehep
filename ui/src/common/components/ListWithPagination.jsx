import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { List, Pagination } from 'antd';
import Immutable from 'immutable';
import classNames from 'classnames';

const GRID_CONFIG = {
  gutter: 16,
  xs: 1,
  sm: 2,
  lg: 3,
  xl: 4,
};

/**
 * Only displays given page items at once and pagination ui
 * does not paginate.
 * Use `ClientPaginatedList` for complete solution, or implement server one like: `CitationList`
 */
class ListWithPagination extends Component {
  static getPaginationRangeInfo(total, range) {
    return `${range[0]}-${range[1]} of ${total}`;
  }

  renderPagination() {
    const { pageSize, loading, total, page, onPageChange, grid } = this.props;
    return (
      <Pagination
        className={classNames({ 'ant-col-24': grid })}
        hideOnSinglePage
        current={page}
        onChange={onPageChange}
        total={total}
        pageSize={pageSize}
        loading={loading}
        showTotal={ListWithPagination.getPaginationRangeInfo}
      />
    );
  }

  // wrap `
  render() {
    const { renderItem, title, pageItems, page, pageSize, grid } = this.props;
    return (
      <List
        header={title}
        footer={this.renderPagination()}
        grid={grid ? GRID_CONFIG : undefined}
      >
        <div
          className={
            classNames({
              'ant-row-flex': grid,
            }) /* to workaround ant-design/issues/14407 */
          }
        >
          {pageItems.map((item, index) =>
            renderItem(item, (page - 1) * pageSize + index)
          )}
        </div>
      </List>
    );
  }
}

ListWithPagination.propTypes = {
  total: PropTypes.number.isRequired,
  pageItems: PropTypes.instanceOf(Immutable.List).isRequired,
  renderItem: PropTypes.func.isRequired,
  onPageChange: PropTypes.func.isRequired,
  pageSize: PropTypes.number.isRequired,
  page: PropTypes.number.isRequired,
  title: PropTypes.node,
  loading: PropTypes.bool,
  grid: PropTypes.bool,
};

ListWithPagination.defaultProps = {
  title: null,
  loading: false,
};

export default ListWithPagination;
