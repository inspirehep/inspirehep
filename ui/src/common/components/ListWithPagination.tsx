import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { List, Pagination } from 'antd';
import Immutable from 'immutable';
import classNames from 'classnames';

const GRID_CONFIG = {
  gutter: 16,
  xs: 1,
  sm: 2,
  lg: 3,
  xl: 4,
};

const PAGE_SIZE_OPTIONS = ['25', '50', '100', '250'];

/**
 * Only displays given page items at once and pagination ui
 * does not paginate.
 * Use `ClientPaginatedList` for complete solution, or implement server one like: `CitationList`
 */
class ListWithPagination extends Component {
  static getPaginationRangeInfo(total: any, range: any) {
    return `${range[0]}-${range[1]} of ${total}`;
  }

  constructor(props: any) {
    super(props);

    this.renderItem = this.renderItem.bind(this);
  }

  renderPagination() {
    const {
      // @ts-expect-error ts-migrate(2339) FIXME: Property 'pageSize' does not exist on type 'Readon... Remove this comment to see the full error message
      pageSize,
      // @ts-expect-error ts-migrate(2339) FIXME: Property 'loading' does not exist on type 'Readonl... Remove this comment to see the full error message
      loading,
      // @ts-expect-error ts-migrate(2339) FIXME: Property 'total' does not exist on type 'Readonly<... Remove this comment to see the full error message
      total,
      // @ts-expect-error ts-migrate(2339) FIXME: Property 'page' does not exist on type 'Readonly<{... Remove this comment to see the full error message
      page,
      // @ts-expect-error ts-migrate(2339) FIXME: Property 'onPageChange' does not exist on type 'Re... Remove this comment to see the full error message
      onPageChange,
      // @ts-expect-error ts-migrate(2339) FIXME: Property 'grid' does not exist on type 'Readonly<{... Remove this comment to see the full error message
      grid,
      // @ts-expect-error ts-migrate(2339) FIXME: Property 'onSizeChange' does not exist on type 'Re... Remove this comment to see the full error message
      onSizeChange,
    } = this.props;
    return (
      <Pagination
        className={classNames({ 'ant-col-24': grid })}
        hideOnSinglePage
        current={page}
        onChange={onPageChange}
        total={total}
        pageSize={pageSize}
        // @ts-expect-error ts-migrate(2769) FIXME: No overload matches this call.
        loading={loading}
        showTotal={ListWithPagination.getPaginationRangeInfo}
        onShowSizeChange={onSizeChange}
        showSizeChanger={onSizeChange != null}
        pageSizeOptions={PAGE_SIZE_OPTIONS}
        responsive
      />
    );
  }

  renderItem(item: any, index: any) {
    // @ts-expect-error ts-migrate(2339) FIXME: Property 'renderItem' does not exist on type 'Read... Remove this comment to see the full error message
    const { renderItem, pageSize, page } = this.props;
    const absoluteIndex = (page - 1) * pageSize + index;
    return renderItem(item, absoluteIndex);
  }

  render() {
    // @ts-expect-error ts-migrate(2339) FIXME: Property 'title' does not exist on type 'Readonly<... Remove this comment to see the full error message
    const { title, grid, pageItems } = this.props;
    return (
      <List
        header={title}
        footer={this.renderPagination()}
        grid={grid ? GRID_CONFIG : undefined}
        dataSource={pageItems}
        renderItem={this.renderItem}
        data-test-id="pagination-list"
      />
    );
  }
}

// @ts-expect-error ts-migrate(2339) FIXME: Property 'propTypes' does not exist on type 'typeo... Remove this comment to see the full error message
ListWithPagination.propTypes = {
  total: PropTypes.number.isRequired,
  // @ts-expect-error ts-migrate(2345) FIXME: Argument of type 'typeof List' is not assignable t... Remove this comment to see the full error message
  pageItems: PropTypes.instanceOf(Immutable.List).isRequired,
  renderItem: PropTypes.func.isRequired,
  onPageChange: PropTypes.func.isRequired,
  onSizeChange: PropTypes.func,
  pageSize: PropTypes.number,
  page: PropTypes.number,
  title: PropTypes.node,
  loading: PropTypes.bool,
  grid: PropTypes.bool,
};

// @ts-expect-error ts-migrate(2339) FIXME: Property 'defaultProps' does not exist on type 'ty... Remove this comment to see the full error message
ListWithPagination.defaultProps = {
  title: null,
  loading: false,
  page: 1,
  pageSize: 25,
};

export default ListWithPagination;
