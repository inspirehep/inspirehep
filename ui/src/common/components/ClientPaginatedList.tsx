import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { List } from 'immutable';
import ListWithPagination from './ListWithPagination';

class ClientPaginatedList extends Component {
  static getDerivedStateFromProps(nextProps, prevState) {
    const { items, pageSize } = nextProps;
    const { prevItems, prevPageSize } = prevState;

    if (items === prevItems && pageSize === prevPageSize) {
      return prevState;
    }

    const { page } = prevState;
    const pageItems = ClientPaginatedList.getPageItems(items, page, pageSize);
    return {
      ...prevState,
      prevItems: items,
      prevPageSize: pageSize,
      pageItems,
      total: items.size,
    };
  }

  static getPageItems(items, page, pageSize) {
    const endIndex = page * pageSize;
    const startIndex = endIndex - pageSize;
    return items.slice(startIndex, endIndex);
  }

  constructor(props) {
    super(props);
    this.onPageChange = this.onPageChange.bind(this);

    this.state = {
      page: 1,
    };
  }

  onPageChange(page) {
    const { items, pageSize } = this.props;
    const pageItems = ClientPaginatedList.getPageItems(items, page, pageSize);
    this.setState({
      pageItems,
      page,
    });
  }

  render() {
    const { renderItem, pageSize, title, items, loading, grid } = this.props;
    const { pageItems, total, page } = this.state;
    return (
      items.size > 0 && (
        <ListWithPagination
          title={title}
          renderItem={renderItem}
          pageItems={pageItems}
          pageSize={pageSize}
          page={page}
          onPageChange={this.onPageChange}
          total={total}
          loading={loading}
          grid={grid}
        />
      )
    );
  }
}

ClientPaginatedList.propTypes = {
  title: PropTypes.node,
  items: PropTypes.instanceOf(List),
  renderItem: PropTypes.func.isRequired,
  loading: PropTypes.bool,
  pageSize: PropTypes.number,
  grid: PropTypes.bool,
};

ClientPaginatedList.defaultProps = {
  items: List(),
  title: null,
  loading: false,
  pageSize: 25,
};

export default ClientPaginatedList;
