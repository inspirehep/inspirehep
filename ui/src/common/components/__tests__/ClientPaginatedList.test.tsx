import React from 'react';
import { shallow } from 'enzyme';
import { fromJS, Range } from 'immutable';
import { List } from 'antd';

import ClientPaginatedList from '../ClientPaginatedList';
import ListWithPagination from '../ListWithPagination';

describe('ClientPaginatedList', () => {
  it('renders first page', () => {
    const items = Range(1, 100).toList();
    const wrapper = shallow(
      <ClientPaginatedList
        items={items}
        renderItem={item => <List.Item key={item}>{item}</List.Item>}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders first page with custom pageSize', () => {
    const items = Range(1, 100).toList();
    const wrapper = shallow(
      <ClientPaginatedList
        items={items}
        pageSize={10}
        renderItem={item => <List.Item key={item}>{item}</List.Item>}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders as loading if set', () => {
    const items = Range(1, 100).toList();
    const wrapper = shallow(
      <ClientPaginatedList
        loading
        items={items}
        renderItem={item => <List.Item key={item}>{item}</List.Item>}
        pageSize={10}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders in grid mode', () => {
    const items = Range(1, 100).toList();
    const wrapper = shallow(
      <ClientPaginatedList
        loading
        items={items}
        renderItem={item => <List.Item key={item}>{item}</List.Item>}
        pageSize={10}
        grid
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the new page on page change', () => {
    const items = Range(1, 100).toList();
    const page = 2;
    const wrapper = shallow(
      <ClientPaginatedList
        items={items}
        renderItem={item => <List.Item key={item}>{item}</List.Item>}
        pageSize={10}
      />
    );
    const onPageChange = wrapper.find(ListWithPagination).prop('onPageChange');
    onPageChange(page);
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });

  it('does not render at all if empty', () => {
    const items = fromJS([]);
    const wrapper = shallow(
      <ClientPaginatedList
        items={items}
        renderItem={item => <List.Item key={item}>{item}</List.Item>}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  describe('getPageItems', () => {
    it('returns items for a page', () => {
      const items = fromJS([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
      const page = 2;
      const pageSize = 3;
      const expected = fromJS([4, 5, 6]);
      const result = ClientPaginatedList.getPageItems(items, page, pageSize);
      expect(result).toEqual(expected);
    });

    it('returns items for the first page', () => {
      const items = fromJS([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
      const page = 1;
      const pageSize = 3;
      const expected = fromJS([1, 2, 3]);
      const result = ClientPaginatedList.getPageItems(items, page, pageSize);
      expect(result).toEqual(expected);
    });

    it('returns only remaining items for the last page', () => {
      const items = fromJS([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
      const page = 4;
      const pageSize = 3;
      const expected = fromJS([10]);
      const result = ClientPaginatedList.getPageItems(items, page, pageSize);
      expect(result).toEqual(expected);
    });
  });
});
