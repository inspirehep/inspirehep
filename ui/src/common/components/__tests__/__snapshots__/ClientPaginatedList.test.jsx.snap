// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientPaginatedList does not render at all if empty 1`] = `""`;

exports[`ClientPaginatedList renders as loading if set 1`] = `
<ListWithPagination
  loading={true}
  onPageChange={[Function]}
  page={1}
  pageItems={
    Immutable.List [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
    ]
  }
  pageSize={10}
  renderItem={[Function]}
  title={null}
  total={99}
/>
`;

exports[`ClientPaginatedList renders first page 1`] = `
<ListWithPagination
  loading={false}
  onPageChange={[Function]}
  page={1}
  pageItems={
    Immutable.List [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14,
      15,
      16,
      17,
      18,
      19,
      20,
      21,
      22,
      23,
      24,
      25,
    ]
  }
  pageSize={25}
  renderItem={[Function]}
  title={null}
  total={99}
/>
`;

exports[`ClientPaginatedList renders first page with custom pageSize 1`] = `
<ListWithPagination
  loading={false}
  onPageChange={[Function]}
  page={1}
  pageItems={
    Immutable.List [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
    ]
  }
  pageSize={10}
  renderItem={[Function]}
  title={null}
  total={99}
/>
`;

exports[`ClientPaginatedList renders in grid mode 1`] = `
<ListWithPagination
  grid={true}
  loading={true}
  onPageChange={[Function]}
  page={1}
  pageItems={
    Immutable.List [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
    ]
  }
  pageSize={10}
  renderItem={[Function]}
  title={null}
  total={99}
/>
`;

exports[`ClientPaginatedList renders the new page on page change 1`] = `
<ListWithPagination
  loading={false}
  onPageChange={[Function]}
  page={2}
  pageItems={
    Immutable.List [
      11,
      12,
      13,
      14,
      15,
      16,
      17,
      18,
      19,
      20,
    ]
  }
  pageSize={10}
  renderItem={[Function]}
  title={null}
  total={99}
/>
`;
